src/gatherers/critical-network-chains.js

```diff
@@ -0,0 +1,149 @@
+  }
+
+  getCriticalChains(networkRecords) {
+    // Drop the first request because it's uninteresting - it's the page html
+    // and always critical. No point including it in every request
+    const criticalRequests = networkRecords.slice(1).filter(
+      req => includes(this.criticalPriorities, req.initialPriority()));
+
+    // Build a map of requestID -> Node.
+    const requestIdToNodes = new Map();
+    for (let request of criticalRequests) {
+      const requestNode = new RequestNode(request, null);
+      requestIdToNodes.set(requestNode.requestId, requestNode);
+    }
+
+    // Connect the parents and children
+    for (let request of criticalRequests) {
+      if (request.initiatorRequest()) {
+        const parentRequestId = request.initiatorRequest().requestId;
+        const childNode = requestIdToNodes.get(request.requestId);
+        const parentNode = requestIdToNodes.get(parentRequestId);
+        if (childNode && parentNode) {
+          // Both child and parent must be critical
+          // TODO: We may need handle redirects carefully. Investigate
+          childNode.setParent(parentNode);
+          parentNode.addChild(childNode);
+        }
+      }
+    }
+
+    const nodesList = [...requestIdToNodes.values()];
+    const orphanNodes = nodesList.filter(node => node.parent === null);
+    const nodeChains = flatten(orphanNodes.map(
+      node => this._getChainsDFS(node)));
+    const requestChains = nodeChains.map(chain => chain.map(
+      node => node.request));
+    return requestChains;
+  }
+
+  postProfiling(options, tracingData) {
+    const chains = this.getCriticalChains(tracingData.networkRecords);
+
+    if (options.flags.useNetDepGraph) {
+      // There logs are here so we can test this gatherer
+      // Will be removed when we have a way to surface them in the report
+      const nonTrivialChains = chains.filter(chain => chain.length > 1);
+
+      // Note: Approximately,
+      // startTime: time when request was dispatched
+      // responseReceivedTime: either time to first byte, or time of receiving
+      //  the end of response headers
+      // endTime: time when response loading finished
+      const debuggingData = nonTrivialChains.map(chain => ({
+        urls: chain.map(request => request._url),
+        totalRequests: chain.length,
+        times: chain.map(request => ({
+          startTime: request.startTime,
+          endTime: request.endTime,
+          responseReceivedTime: request.responseReceivedTime
+        })),
+        totalTimeBetweenBeginAndEnd:
+          (chain[chain.length - 1].endTime - chain[0].startTime),
+        totalLoadingTime: chain.reduce((acc, req) =>
+          acc + (req.endTime - req.responseReceivedTime), 0)
+      }));
+      log.log('info', 'cricital chains', JSON.stringify(debuggingData));
+      log.log('info', 'lengths of critical chains', debuggingData.map(d => d.totalRequests));
+    }
+    return {CriticalNetworkChains: chains};
+  }
+
```

Comment:
type docs would help here (and above) for clarity, even if not checked. Hard to wade through all the flattens :) `@return {!Array<!Array<!RequestNode>>}`?